src/coreclr/nativeaot/Runtime/i386/MiscStubs.asm

```diff
@@ -39,4 +39,145 @@ ProbeLoop:
 
+;; *********************************************************************/
+;; LLsh - long shift left
+;;
+;; Purpose:
+;;    Does a Long Shift Left (signed and unsigned are identical)
+;;    Shifts a long left any number of bits.
+;;
+;; Entry:
+;;    EDX:EAX - long value to be shifted
+;;        ECX - number of bits to shift by
+;;
+;; Exit:
+;;    EDX:EAX - shifted value
+;;
+;; NOTE: Adapted from JIT_LLsh in CoreCLR
+;;
+RhpLLsh PROC public
+    ;; Reduce shift amount mod 64
+    and     ecx, 63
+
+    cmp     ecx, 32
+    jae     LLshMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shld    edx, eax, cl
+    shl     eax, cl
+    ret
+
+LLshMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     edx, eax
+    xor     eax, eax
+    shl     edx, cl
+    ret
+RhpLLsh ENDP
+
+;; *********************************************************************/
+;; LRsh - long shift right
+;;
+;; Purpose:
+;;    Does a signed Long Shift Right
+;;    Shifts a long right any number of bits.
+;;
+;; Entry:
+;;    EDX:EAX - long value to be shifted
+;;        ECX - number of bits to shift by
+;;
+;; Exit:
+;;    EDX:EAX - shifted value
+;;
+;; NOTE: Adapted from JIT_LRsh in CoreCLR
+;;
+RhpLRsh PROC public
+    ;; Reduce shift amount mod 64
+    and     ecx, 63
+
+    cmp     ecx, 32
+    jae     LRshMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shrd    eax, edx, cl
+    sar     edx, cl
+    ret
+
+LRshMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     eax, edx
+    sar     edx, 31
+    sar     eax, cl
+    ret
+RhpLRsh ENDP
+
+;; *********************************************************************/
+;; LRsz
+;;
+;; Purpose:
+;;    Does a unsigned Long Shift Right
+;;    Shifts a long right any number of bits.
+;;
+;; Entry:
+;;    EDX:EAX - long value to be shifted
+;;        ECX - number of bits to shift by
+;;
+;; Exit:
+;;    EDX:EAX - shifted value
+;;
+;; NOTE: Adapted from JIT_LRsz in CoreCLR
+;;
+RhpLRsz PROC public
+    ;; Reduce shift amount mod 64
+    and     ecx, 63
+
+    cmp     ecx, 32
+    jae     LRszMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shrd    eax, edx, cl
+    shr     edx, cl
+    ret
+
+LRszMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     eax, edx
+    xor     edx, edx
+    shr     eax, cl
+    ret
+RhpLRsz ENDP
+
+;; *********************************************************************/
+;; LMul
+;;
+;; Purpose:
+;;    Does a long multiply (same for signed/unsigned)
+;;
+;; Entry:
+;;    Parameters are passed on the stack:
+;;    1st pushed: multiplier (QWORD)
+;;    2nd pushed: multiplicand (QWORD)
+;;
+;; Exit:
+;;    EDX:EAX - product of multiplier and multiplicand
+;;
+;; NOTE: Adapted from clang generated code
+;;
+RhpLMul PROC public
+    push    esi
```

Comment:
CoreCLR C/C++ version of this helper is special cased for small positive numbers. It is reasonable to expect that small positive numbers are very common, and it is worth the special casing, somebody must have done that intentionally. Can we use C++ version of this helper that matches the existing CoreCLR one for this one too?

Reply:
The current NativeAOT helper in MathHelpers.cpp is
```
EXTERN_C int64_t F_CALL_CONV RhpLMul(int64_t i, int64_t j)
{
    return i * j;
}
```

On MSVC it compiles to call to `__allmul` which in my tests was slower than this assembly version (possibly due to additional call overhead). The assembly code matches clang output with optimizations (+ calling convention difference on ret). While that version may not have a special code path for small positive numbers it does use only two `imul`s instead of three `mul`s + conditional jump that the CoreCLR x86 ASM helper uses in the worst case.

Changing to the CoreCLR's C version would not match what CoreCLR uses today on x86. Using the CoreCLR x86 assembly helper would match CoreCLR but I have strong suspicion that the slow code path is actually worse than this version. Using current NativeAOT C helper also doesn't seem to be a win when compiled with MSVC, and it's not any better/worse when compiled with Clang.

Reply:
In summary, we cannot move this to C without a performance penalty. We can either keep this version (which is likely optimal for the general case), use the CoreCLR assembly version (which almost certainly is worse than this one for the "hard" case), or try to blend the two assembly versions together to add the optimization.

Reply:
I incorrectly stated that there's different number of `[i]mul`s between this version and the CoreCLR assembly version. I crossed out the mistake. The conclusion remains.

Reply:
I copied the CoreCLR assembly version which has the 32x32->64 optimization.

Reply:
I understand that there are tradeoffs and that the current CoreCLR version may not be optimal on current hardware. 

My point is that both regular CoreCLR and native AOT should be identical where possible. It is what makes native AOT viable. Ideally, the code would be physically shared. Until we get there, the copies of the code should be kept as similar as possible and needless differences should be avoided.

Reply:
In general I agree with you about keeping CoreCLR and NativeAOT identical where possible and maximizing the code sharing. However, when adding new platform to NativeAOT I ultimately prefer sharing code with existing NativeAOT code / platforms over copying CoreCLR code. For the assembly helpers there were two additional factors. Firstly, CoreCLR implements some of the operations in multiple variations for different hardware baselines and dynamically switches them at runtime. That concept doesn't easily translate to NativeAOT and requires additional complexity that is not desirable for new platform bringup (ie. main goal is having a correct baseline that passes tests, not optimal baseline that happens to pass the tests on my AVX512 HW). Secondly, at the time I wrote the code there were in-flight PRs that would remove and refactor parts of these helpers so I tried to get something that passes as minimal viable prototype without spending too much time on something that others were already trying to optimize. 

In case of `RhpLMul` sharing the existing NativeAOT version was not viable (both for performance reasons and lack of the `FCIMPL1_L` macro magic at that time), and neither was reusing the C version of CoreCLR code (which happens to be even worse when compiled with MSVC). That left me with the choice of copying 30 lines of assembly from CoreCLR or getting something reasonably fast and working with 11 lines of Clang generated assembly. At first glance the Clang version seemed like a clear win (and unlike the long->double conversions this doesn't easily benefit from optimized instruction sets on new HW).

After a more careful comparison of the CoreCLR assembly code I agree that it's fine to share the assembly version which is what I committed yesterday.

Reply:
I have not realized that CoreCLR has handwritten assembly for the 64-bit multiplication on Windows x86. Looks good! Thanks.